programs/cltester/dwarf.cpp

```diff
@@ -523,6 +592,146 @@ namespace dwarf
       }
    }
 
+   // wasm_addr is relative to beginning of file
+   std::optional<uint32_t> get_wasm_fn(const info& info, uint32_t wasm_addr)
+   {
+      auto it = std::upper_bound(info.wasm_fns.begin(), info.wasm_fns.end(), wasm_addr,
+                                 [](auto a, const auto& b) { return a < b.size_pos; });
+      if (it == info.wasm_fns.begin())
+         return {};
+      --it;
+      if (it->size_pos <= wasm_addr && wasm_addr < it->end_pos)
+         return &*it - &info.wasm_fns[0];
+      return {};
+   }
+
+   std::optional<std::pair<uint64_t, uint64_t>> get_addr_range(
+       const info& info,
+       const std::vector<jit_fn_loc>& fn_locs,
+       const std::vector<jit_instr_loc>& instr_locs,
+       const void* code_start,
+       uint32_t begin,
+       uint32_t end)
+   {
+      // TODO: cuts off a range which ends at the wasm's end
+      auto it1 =
+          std::lower_bound(instr_locs.begin(), instr_locs.end(), info.wasm_code_offset + begin,
+                           [](const auto& a, auto b) { return a.wasm_addr < b; });
+      auto it2 = std::lower_bound(instr_locs.begin(), instr_locs.end(), info.wasm_code_offset + end,
+                                  [](const auto& a, auto b) { return a.wasm_addr < b; });
+      if (it1 < it2 && it2 != instr_locs.end())
+      {
+         return std::pair{uint64_t((char*)code_start + it1->code_offset),
+                          uint64_t((char*)code_start + it2->code_offset)};
+      }
+      return {};
+   }
+
+   template <typename S>
+   void write_line_program(const info& info,
+                           const std::vector<jit_fn_loc>& fn_locs,
+                           const std::vector<jit_instr_loc>& instr_locs,
+                           const void* code_start,
+                           S& s)
+   {
+      std::optional<uint64_t> address = 0;
```

Comment:
Why is `address` an `optional`?  I don't see any way that it can be set to `nullopt`.

Reply:
changed